### DISCUSSION TITLE
Solving Eigenvalue problem for a square plate with a circular hole

### DISCUSSION THREAD

**Discussion by adhithya96** [idaholab/moose#27072] [Q&A Modules: Solid mechanics]
I am trying to solve the eigenvalue problem for a square plate with a hole using the XFEM and level set modules. I checked out a few examples on the MOOSE repository for both modules and setup an input file for the same. I am getting some output but I am unable to see the circular hole or any change in eigenvalues. The following is the input file I am using
```
[GlobalParams]
    displacements = 'ux uy'
[]  

[XFEM]
    geometric_cut_userobjects = 'cut_mesh'
    qrule = volfrac
    output_cut_plane = true
[]
  

[Mesh]
    [First]
        type = FileMeshGenerator
        file = square.msh
    []
[]

[Modules/TensorMechanics/Master]
    [./all]
      strain = SMALL
      planar_formulation = PLANE_STRAIN
      add_variables = true
      generate_output = 'stress_xx stress_xy stress_yy stress_zz strain_xx strain_xy strain_yy strain_zz'
    [../]
[]
  

[UserObjects]
    [cut_mesh]
      type = InterfaceMeshCut2DUserObject
      mesh_file = circle_surface.e
      interface_velocity_function = vel_func
      #heal_always = true
    []
[]

[Functions]
    [vel_func]
      type = ConstantFunction
      value = 0
    []
[]

[AuxVariables]
    [ls]
      order = FIRST
      family = LAGRANGE
    []
[]  

[Kernels]
    [eigen1]
        type = CoefReaction
        variable = ux
        coefficient = -1.0
        extra_vector_tags = 'eigen'
    []

    [eigen2]
        type = CoefReaction
        variable = uy
        coefficient = -1.0
        extra_vector_tags = 'eigen'
    []
[]

[AuxKernels]
    [ls]
      type = MeshCutLevelSetAux
      mesh_cut_user_object = cut_mesh
      variable = ls
      execute_on = 'TIMESTEP_BEGIN'
    []
  []
  

[BCs]
    [fixed1]
        type = DirichletBC
        variable = ux
        boundary = 'boundary'
        value = 0
    []
    [fixed2]
        type = DirichletBC
        variable = uy
        boundary = 'boundary'
        value = 0
    []

[]

[Materials]
    [./linear_stress]
        type = ComputeLinearElasticStress
    [../]
    [./elasticity_tensor]
        type = ComputeIsotropicElasticityTensor
        poissons_ratio = 0.3
        youngs_modulus = 1e10
    [../]
[]

[VectorPostprocessors]
    [eigen]
      type = Eigenvalues
      inverse_eigenvalue = true
    []
[]
  
#[Postprocessors]
#    [fluxintegral]
#      type = ElementIntegralVariablePostprocessor
#      variable = u
#      execute_on = linear
#    []
#[]

[Problem]
    type = EigenProblem
[]

#Linear  eigenvalue solver
[Executioner]
    type = Eigenvalue
    solve_type = arnoldi
    n_eigen_pairs = 4
[]

[Outputs]
    file_base =  eigen_circularhole_out
    exodus =  true
    csv = true
[]
```

**Comment by GiudGiud** [comment]
Hello

I just run your case with a couple meshes that have the same name that I found in the moose folder and I immediately hit this error:
```
*** ERROR ***
The following error occurred in the object "cut_mesh", of type "InterfaceMeshCut2DUserObject".

cutFragmentByGeometry not yet implemented for InterfaceMeshCut2DUserObject
```

so it seems there is some development work to do on this first.

Guillaume

**Reply by adhithya96** [reply]
Oh damn! Will look into the codes and get back to you with more questions. The moose examples have bimaterial interface but none with a hole. I was thinking I'll use it and use zero Young's modulus and poisson's ratio and try but yeah will see what I can do.

**Reply by adhithya96** [reply]
I tried two things 
1. I used the ```LevelSetCutUserObject``` in MOOSE to specify the circular hole using equation ```$x^2  + y^2 - 25 = 0 $```. I used the same square mesh. Some of the input files given as example  had a bimaterial interface. So I used it to change the materials properties to 0 (very close to zero) to see if it'll work. When I use ADTypeproperty, I got mismatch between AD and non-AD property.  When I changed it to nonAD type material  and strains I am getting this error
```
The following error occurred in the object "MOOSE Problem", of type "EigenProblem".

Material property 'Jacobian_mult', requested by 'TM_TensorMechanics0' is not defined on block 1
Material property 'Jacobian_mult', requested by 'TM_TensorMechanics1' is not defined on block 1
```
How can one mention subdomain when there  is a square and hole in the center ?

2. I copied one of the examples  of bimaterial from xfem module and changed the executioner and kernel to see if the eigenvalue problem can be coupled with xfem. Ran into this error
```
The following error occurred in the object "MOOSE Problem", of type "EigenProblem".

A libMesh::PetscSolverException was raised during FEProblemBase::computeJacobianTags
New nonzero at (0,0) caused a malloc
Use MatSetOption(A, MAT_NEW_NONZERO_ALLOCATION_ERR, PETSC_FALSE) to turn off this check
```
I have  added the input files in the github [repo](https://github.com/adhithya96/EigenSolverMOOSE) in the files titled ```eigen2d_xfem_trial2.i``` and ```eigen2d_xfem_trial3.i```

**Reply by GiudGiud** [reply]
> How can one mention subdomain when there is a square and hole in the center ?

the hole is not meshed right? So you dont need to worry about defining a material there
what is block 1 in the mesh?

> I copied one of the examples of bimaterial from xfem module and changed the executioner and kernel to see if the eigenvalue problem can be coupled with xfem. Ran into this error

see other thread. You can allow these re-allocations, which are due to XFEM

**Reply by adhithya96** [reply]
There is no hole. When the level set module represents a circle, the phantom nodes are created differentiating between the two domains and hence two materials. All the other examples I found in xfem rated to solid mechanics are related to crack propagation. This is the only thing I thought might work. 
Will check out the other threads

**Reply by GiudGiud** [reply]
ok then define the materials in all the blocks that the equations are defined on. If some equations should not be defined in the hole, then block restrict the kernels or the variables
